File: python/dp/graph/$787_CheapestFlightsWithinKStops.py
from collections import defaultdict
from functools import lru_cache
from typing import List


def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
    connect = defaultdict(dict)
    inf = 1e9+7
    for a, b, c in flights:  # 创建邻接矩阵
        connect[a][b] = c

    @lru_cache(None)
    def dfs(city, remain):
        if city == dst:  # 到达目标城市，返回零
            return 0
        if not remain:  # 不是终点也不能再动了，返回inf
            return inf
        remain -= 1  # 剩下城市数量减一
        ans = inf
        for nxt in connect[city]:
            ans = min(ans, dfs(nxt, remain) + connect[city][nxt])
        return ans

    res = dfs(src, k + 1)
    return res if res != inf else -1

File: java/src/main/java/cn/leetcode/dp/graph/$787_CheapestFlightsWithinKStops.java
package cn.leetcode.dp.graph;

import java.util.Arrays;

public class $787_CheapestFlightsWithinKStops {
    final int INF = 10000 * 101 + 1;
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {

        int[][] f = new int[k + 2][n];
        for (int i = 0; i < k + 2; ++i) {
            Arrays.fill(f[i], INF);
        }
        f[0][src] = 0;
        for (int t = 1; t <= k + 1; ++t) {  // 第1次航班到第k+1次航班
            for (int[] flight : flights) {
                int j = flight[0], i = flight[1], cost = flight[2];
                f[t][i] = Math.min(f[t][i], f[t - 1][j] + cost);
            }
        }
        int ans = INF;
        for (int t = 1; t <= k + 1; ++t) {
            ans = Math.min(ans, f[t][dst]);
        }
        return ans == INF ? -1 : ans;
    }

    public int findCheapestPrice1d(int n, int[][] flights, int src, int dst, int k) {
        int[] f = new int[n];
        Arrays.fill(f, INF);
        f[src] = 0;
        int ans = INF;
        for (int t = 1; t <= k + 1; ++t) {
            int[] g = new int[n];
            Arrays.fill(g, INF);
            for (int[] flight : flights) {
                int j = flight[0], i = flight[1], cost = flight[2];
                g[i] = Math.min(g[i], f[j] + cost);
            }
            f = g;
            ans = Math.min(ans, f[dst]);
        }
        return ans == INF ? -1 : ans;
    }


    public static void main(String[] args) {
        int n = 3;
        int[][] edges = {{0,1,100},{1,2,100},{0,2,500}};
        int src = 0, dst = 2, k = 1;
        int res = new $787_CheapestFlightsWithinKStops().findCheapestPrice(n, edges, src, dst, k);
        System.out.println(res);
    }
}
